nupic/research/frameworks/dendrites/utils.py

```diff
@@ -19,78 +19,251 @@
+                            dim_context) where each 2D tensor gives a batch of context
+                            vectors from the same category
+    :param selection_criterion: the criterion for selecting which dendrites become
+                                active; either "regular" (for `GatingDendriticLayer`)
+                                or "absolute" (for `AbsoluteMaxGatingDendriticLayer`)
+    """
+    overlap_matrix = dendrite_overlap_matrix(dendrite_segments, context_vectors,
+                                             selection_criterion)
+
+    num_units, num_categories, _ = overlap_matrix.size()
+
+    # The overlap score is simply the average of the off-diagonal entries of the
+    # overlap matrix; in the ideal case with no dendrite overlap, the overlap matrix is
+    # the identity matrix
+
+    # Since the overlap matrix is symmetric, we only consider the lower half, excluding
+    # the diagonal entries since they are all guaranteed to be 1
+
+    # Mask for a lower triangular matrix
+    ltril_mask = torch.ones((num_categories, num_categories))
+    ltril_mask = ltril_mask.tril(diagonal=-1)
+    ltril_mask = ltril_mask.unsqueeze(0).repeat((num_units, 1, 1))
+
+    overlap_score = (overlap_matrix * ltril_mask).sum(dim=(1, 2))
+    overlap_score /= (0.5 * num_categories * (num_categories - 1))
+    return overlap_score
+
+
+def dendrite_duty_cycle(dendrite_segments, context_vectors, selection_criterion):
+    """
+    Returns a 2D torch tensor with shape (number of units, number of dendrites) where
+    entry i, j gives the duty cycle of dendrite j in the set of dendrite segments for
+    unit i. The duty cycle for each dendrite is computed offline.
+
+    :param dendrite_segments: `DendriteSegments` object
+    :param context_vectors: a single 2D torch tensor of context vectors across multiple
+                            classes, or iterable of 2D torch tensors with shape
+                            (num_examples, dim_context) where each 2D tensor gives a
+                            batch of context vectors from the same category
+    :param selection_criterion: the criterion for selecting which dendrites become
+                                active; either "regular" (for `GatingDendriticLayer`)
+                                or "absolute" (for `AbsoluteMaxGatingDendriticLayer`)
+    """
+    if not isinstance(context_vectors, torch.Tensor):
+        context_vectors = torch.cat(context_vectors)
+    num_examples = context_vectors.size(0)
+
+    activations = dendrite_segments(context_vectors)
+    if selection_criterion == "absolute":
+        activations = activations.abs()
+
+    duty_cycle = (activations.max(axis=2, keepdims=True).values == activations)
+    duty_cycle = duty_cycle.sum(axis=0, dtype=torch.float)
+    duty_cycle = duty_cycle / num_examples
+
+    return duty_cycle
+
+
+def online_dendrite_duty_cycle(dendrite_segments, context_vectors, selection_criterion,
```

Comment:
Is this function getting used at all? I don't see any reference to it in `nupic.research`.

Reply:
It's not. I only left it in because I thought we may want to consider duty cycle that's not over the entire learning trajectory, and only the last T samples seen. I think I can just merge `dendrite_duty_cycle` and `online_dendrite_duty_cycle` into 1 function with a parameter that controls how duty cycle values are computed.

Reply:
In general, we should be wary of added code or functionality that's not being currently used. Code that's unused goes untested and quickly becomes outdated. You should save the function on a separate branch, patch, or stash, and reintroduce it as needed.

Reply:
Ok, I have removed this function from the commit.

Reply:
Much appreciated.